src/VkBootstrap.cpp

```diff
@@ -1086,89 +1091,92 @@ PhysicalDeviceSelector::Suitable PhysicalDeviceSelector::is_device_suitable(Phys
+			return PhysicalDevice::Suitable::no;
 		}
 	}
-	if (criteria.require_present && !swapChainAdequate) return Suitable::no;
 
-	if (pd.device_properties.deviceType != static_cast<VkPhysicalDeviceType>(criteria.preferred_type)) {
+	if (pd.properties.deviceType != static_cast<VkPhysicalDeviceType>(criteria.preferred_type)) {
 		if (criteria.allow_any_type)
-			suitable = Suitable::partial;
+			suitable = PhysicalDevice::Suitable::partial;
 		else
-			return Suitable::no;
+			return PhysicalDevice::Suitable::no;
 	}
 
 	bool required_features_supported = detail::supports_features(
-	    pd.device_features, criteria.required_features, pd.extended_features_chain, criteria.extended_features_chain);
-	if (!required_features_supported) return Suitable::no;
+	    pd.features, criteria.required_features, pd.extended_features_chain, criteria.extended_features_chain);
+	if (!required_features_supported) return PhysicalDevice::Suitable::no;
 
 	bool has_required_memory = false;
 	bool has_preferred_memory = false;
-	for (uint32_t i = 0; i < pd.mem_properties.memoryHeapCount; i++) {
-		if (pd.mem_properties.memoryHeaps[i].flags & VK_MEMORY_HEAP_DEVICE_LOCAL_BIT) {
-			if (pd.mem_properties.memoryHeaps[i].size > criteria.required_mem_size) {
+	for (uint32_t i = 0; i < pd.memory_properties.memoryHeapCount; i++) {
+		if (pd.memory_properties.memoryHeaps[i].flags & VK_MEMORY_HEAP_DEVICE_LOCAL_BIT) {
+			if (pd.memory_properties.memoryHeaps[i].size > criteria.required_mem_size) {
 				has_required_memory = true;
 			}
-			if (pd.mem_properties.memoryHeaps[i].size > criteria.desired_mem_size) {
+			if (pd.memory_properties.memoryHeaps[i].size > criteria.desired_mem_size) {
 				has_preferred_memory = true;
 			}
 		}
 	}
-	if (!has_required_memory) return Suitable::no;
-	if (!has_preferred_memory) suitable = Suitable::partial;
+	if (!has_required_memory) return PhysicalDevice::Suitable::no;
+	if (!has_preferred_memory) suitable = PhysicalDevice::Suitable::partial;
 
 	return suitable;
 }
+// delegate construction to the one with an explicit surface parameter
+PhysicalDeviceSelector::PhysicalDeviceSelector(Instance const& instance)
+: PhysicalDeviceSelector(instance, VK_NULL_HANDLE) {}
 
-PhysicalDeviceSelector::PhysicalDeviceSelector(Instance const& instance) {
+PhysicalDeviceSelector::PhysicalDeviceSelector(Instance const& instance, VkSurfaceKHR surface) {
 	instance_info.instance = instance.instance;
-	instance_info.headless = instance.headless;
 	instance_info.version = instance.instance_version;
 	instance_info.supports_properties2_ext = instance.supports_properties2_ext;
+	instance_info.surface = surface;
 	criteria.require_present = !instance.headless;
 	criteria.required_version = instance.instance_version;
 	criteria.desired_version = instance.instance_version;
 }
 
-detail::Result<PhysicalDevice> PhysicalDeviceSelector::select() const {
-
+detail::Result<std::vector<PhysicalDevice>> PhysicalDeviceSelector::select_impl(DeviceSelectionMode selection) const {
```

Comment:
No, because any more headers is undesired. std::set is not commonly included and i'd have to define the key, when its really the drivers responsibility to not return duplicates.